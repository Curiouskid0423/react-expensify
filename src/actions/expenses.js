/**
 * Actions file for expenses.
 * @author Kevin Li
 */

import database from "../firebase/firebase";


/**
 * @desc A synchronous action generator that returns an object, based
 * on the params passed in by startAddExpense.
 * @param expense
 * @return {{type: string, expense: *}}
 */
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

/**
 * @desc startAddExpense bootstraps the addExpense action dispatch.
 * @param description,  amount, createdAt, note
 * @return an async dispatch function call.
 */
export const startAddExpense = (
    { description = "", amount = 0, createdAt = 0, note = "" } = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        const expense = { description, amount, createdAt, note };
        // Returning this database call gives a Promise that may be tested on,
        // but the code can work completely fine even if the call isn't returned.
        return database.ref(`user/${userId}/expense`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}

/**
 * Action for removing expense.
 * @param id
 * @returns {{id: *, type: string}}
 */
export const removeExpense = (id) => ({
        type: "REMOVE_EXPENSE", id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        return database.ref(`user/${userId}/expense/${id}`).set(null).then(() => {
            dispatch(removeExpense(id));
        }, (error) => {
            console.log("Failed at removing an expense:", error);
        });
    }
}


/**
 * Action for editing expense based on provided info.
 * @param id
 * @param editObj
 * @returns {{edit: *, id: *, type: string}}
 */
export const editExpense = (id, editObj) => ({
        type: "EDIT_EXPENSE",
        id,
        edit: editObj
});

export const startEditExpense = (id, editObj = {}) => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
         return database.ref(`user/${userId}/expense/${id}`)
             .update(editObj)
             .then(() => {
             dispatch(editExpense(id, editObj));
         });
    }
}

export const setExpenses = (expenses) => {
    return {
        type: "SET_EXPENSES",
        expenses
    };
}

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const userId = getState().auth.uid;
        return database.ref(`user/${userId}/expense`)
            .once("value")
            .then((snapshot) => {
                const lst = [];
                snapshot.forEach((child) => {
                    lst.push({
                        id: child.key,
                        ...child.val()
                    });
                });
                dispatch(setExpenses(lst));
            });
    }
}