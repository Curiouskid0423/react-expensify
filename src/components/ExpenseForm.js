import React from "react";
import moment from "moment";
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates";

/**
 * ExpenseForm will be reused in Edit Expense and Add Expense.
 * `react-dates` has some outdated dependencies. Beware.
 * @author Kevin Li
 */

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const isEdit = !!props.expense;
        this.state = {
            description: isEdit ? props.expense.description : "",
            amount: isEdit ?
                (props.expense.amount / 100).toString() : "",
            note: isEdit ? props.expense.note : "",
            createdAt: isEdit ? moment(props.expense.createdAt) : moment(),
            calFocused: false,
            error: ""
        }
    }

    // Event listener for description change
    onDescChange = (e) => {
        const description = e.target.value;
        this.setState({ description });
    }

    // Event listener for amount changes, with regex check
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{1,2})?$/)) {
            this.setState({ amount });
        }
    }

    // Event listener for note changes
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState({ note });
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState({ createdAt });
        }
    }

    onAddExpense = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            const error = "Please fill out the form.";
            this.setState({ error });
        } else {
            this.setState({error: ""});
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                { this.state.error && <h4>{ this.state.error }</h4> }
                <form onSubmit = {this.onAddExpense}>
                    <input type = "text" placeholder = "Description"
                           value = {this.state.description} autoFocus = {true}
                           onChange={this.onDescChange}
                    />
                    <input type = "number" placeholder = "Amount" value={this.state.amount}
                           onChange={this.onAmountChange}/>
                    <br />
                    <textarea placeholder = "Optional notes for the expense"
                              value = {this.state.note}
                              onChange={this.onNoteChange}>
                    </textarea>
                    <SingleDatePicker date = {this.state.createdAt}
                                      onDateChange = {this.onDateChange}
                                      focused = {this.state.calFocused}
                                      onFocusChange = {({ focused }) => this.setState({calFocused: focused})}
                                      numberOfMonths = {1} isOutsideRange = {() => false}
                    />
                    <br/>
                    <button type="submit"> Submit </button>
                </form>
            </div>
        )
    }
}