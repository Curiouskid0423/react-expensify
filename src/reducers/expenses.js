/**
 * Expenses reducer file.
 * @author Kevin Li
 */

//TODO: Best practice to keep `reducer` just updating the value, and avoid computations.
//  Do the computations in action generators.
/**
 *  Expense reducer keeps track of the expenses array, i.e. only
 *  modifies the "expenses" portion of  the central state.
 * @param prevState
 * @param action Action to react to
 * @returns {*[]}
 */
const expenseReducer = (prevState = [], action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [
                ...prevState,
                action.expense
            ];
        case "REMOVE_EXPENSE":
            return prevState.filter((item) => item.id !== action.id);
        case "EDIT_EXPENSE":
            return prevState.map((item) => {
                if (item.id === action.id) {
                    item = {
                        ...item,
                        ...action.edit
                    }
                }
                return item;
            });
        default:
            return prevState;
    }
};

export default expenseReducer;