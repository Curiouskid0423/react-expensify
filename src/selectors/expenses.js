/**
 * Selector methods for expenses.
 * @author Kevin Li
 */

/**
 * Getter method for visible expenses based on filters.
 * @param expenses
 * @param filters
 * @returns {*}
 */
export const getVisibleExpenses = (expenses,
                            {text, sortBy, startDate, endDate} = filterDefault) => {
    const unSorted = expenses.filter((item) => {
        const textMatch = item.description.toLowerCase().includes(text.toLowerCase());
        const startMatch = typeof startDate !== "number" || startDate <= item.createdAt;
        const endMatch = typeof endDate !== "number" || endDate >= item.createdAt;
        return textMatch && startMatch && endMatch;
    });
    return sortExpenses(unSorted, sortBy);
}

/**
 * Helper method for sorting expense
 * @param target is the array to be sorted
 * @param sortBy is an indicator of what to sort by
 * @returns a sorted array of expenses
 */
export const sortExpenses = (target, sortBy) => {
    if (sortBy === "amount") {
        return target.sort((a, b) => b.amount - a.amount);
    }
    return target.sort((a, b) => b.createdAt - a.createdAt);
}