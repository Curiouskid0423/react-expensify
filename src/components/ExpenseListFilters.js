import React from "react";
import {connect} from "react-redux";
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
import { DateRangePicker } from "react-dates";

// Export the unconnected version of ExpenseListFilters for testing.
export class ExpenseListFilters extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            calFocused: null
        }
    }

    /**
     * Helper method to update sortBy with select options.
     * @param e is the data from event listener.
     */
    updateSortBy = (e) => {
        const value = e.target.value;
        if (value === "amount") {
            this.props.sortByAmount();
        } else {
            this.props.sortByDate();
        }
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calFocused) => {
        this.setState(() => ({ calFocused }));
    }

    /**
     * Dummy helper method to generate date id for react-dates library
     * @param moment
     * @return an id for the passed in moment object
     */
    getDateId = (moment) => {
        return moment ? moment.toString() : "emptyDate";
    }

    render() {
        return (
            <div>
                <input type = "text" onChange={this.onTextChange}/> &nbsp;
                <select onChange = {this.updateSortBy}>
                    <option value = "date"> Date </option>
                    <option value = "amount"> Amount </option>
                </select>
                <DateRangePicker
                    startDate = { this.props.filters.startDate }
                    startDateId = {this.getDateId(this.props.filters.startDate)}
                    endDate = { this.props.filters.endDate }
                    endDateId = {this.getDateId(this.props.filters.endDate)}
                    onDatesChange = { this.onDatesChange }
                    focusedInput = {this.state.calFocused}
                    onFocusChange = {this.onFocusChange}
                    isOutsideRange = {() => false}
                    numberOfMonths = {1} showClearDates = {true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (dateVal) => dispatch(setStartDate(dateVal)),
        setEndDate: (dateVal) => dispatch(setEndDate(dateVal)),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);