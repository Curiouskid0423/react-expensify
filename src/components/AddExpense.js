import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import {addExpense} from "../actions/expenses";

export class AddExpense extends React.Component {
    constructor(props) {
        super(props);
    }
    // This component can be transformed into a stateless component, but
    // currently being a classic just for the sake of testing.
    onSubmit = (expenseObj) => {
        this.props.dispatchAdder(expenseObj);
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <br/>
                <ExpenseForm onSubmit = {this.onSubmit}/>
            </div>
        );
    }
}

// Use mapDispatchToProps to simplify functions for better testing experience.
const mapDispatchToProps = (dispatch) => {
    return {
        dispatchAdder: (expenseObj) => dispatch(addExpense(expenseObj))
    };
};

// TODO: We can safely use `connect` without `mapStateToProps` if we don't need info from
//  the store other than the function `dispatch()`, in order to update the redux store.
export default connect(undefined, mapDispatchToProps)(AddExpense);