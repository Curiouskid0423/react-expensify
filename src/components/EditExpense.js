import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import {editExpense, removeExpense} from "../actions/expenses";

export class EditExpense extends React.Component {
    constructor(props) {
        super(props);
    }

    onEdit = (expense) => {
        const id = this.props.expense.id;
        this.props.dispatchEdit(id, expense);
        this.props.history.push("/");
    }

    onRemoval = () => {
        this.props.dispatchRemove({ id: this.props.expense.id});
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <ExpenseForm expense = {this.props.expense} onSubmit = {this.onEdit} />
                <button onClick = {this.onRemoval}> Remove </button>
            </div>
        )
    }

}

// match.params.id is implicitly returned by React Component.
// ownProps take in the props from the connected component.
const mapStateToProps = (state, ownProps) => {
    return {
        expense: state.expenses.find((item) => {
            return item.id === ownProps.match.params.id;
        })
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        dispatchEdit: (id, expense) => dispatch(editExpense(id, expense)),
        dispatchRemove: (id) => dispatch(removeExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);