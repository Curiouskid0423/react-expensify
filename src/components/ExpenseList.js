import React from "react";
import { connect } from "react-redux";
import ExpenseItem from "./ExpenseListItem";
import { getVisibleExpenses } from "../selectors/expenses";

/**
 * Helper method to load each expense.
 * @param props
 * @return {*}
 */
const ExpensesLoader = (props) => {
    if (props.expenses.length === 0) {
        return (<h4>No expense is tracked under the given condition.</h4>);
    }
    return props.expenses.map((item) => <ExpenseItem key = {item.id} {...item} />);
}

export const ExpenseList = (props) => {
    return (
        <div>
            <ExpensesLoader expenses = {props.expenses} />
        </div>
    )
}

/**
 * MapStateToProps take in two arguments: state, and ownProps
 * The `state` object could actually be called store (redux)
 * @param state
 * @return a `props` object
 */
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);