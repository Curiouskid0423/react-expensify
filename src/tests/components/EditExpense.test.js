import React from "react";
import { shallow } from "enzyme";
import {EditExpense} from "../../components/EditExpense";
import {expenses as defaultExp} from "../fixtures/expenses";

describe("EditExpense page behaviors.", () => {

    let wrapper, spyEdit, spyRemove, history;
    let selectedExp = defaultExp[1];
    beforeEach(() => {
        spyEdit = jest.fn();
        spyRemove = jest.fn();
        history = { push: jest.fn() }
        wrapper = shallow(<EditExpense
            expense = { selectedExp } dispatchEdit = {spyEdit}
            dispatchRemove = {spyRemove} history = {history}
        />);
    })

    it("should successfully render a default snapshot.", () => {
       expect(wrapper).toMatchSnapshot();
    });

    it("should handle editExpense action.", () => {
        wrapper.find("ExpenseForm").prop("onSubmit")(selectedExp);
        expect(spyEdit).toHaveBeenLastCalledWith(selectedExp.id, selectedExp);
        expect(history.push).toHaveBeenLastCalledWith("/");
    });

    it("should handle removeExpense action.", () => {
       wrapper.find("button").simulate("click");
       expect(spyRemove).toHaveBeenLastCalledWith({id: selectedExp.id });
       expect(history.push).toHaveBeenLastCalledWith("/");
    });

});