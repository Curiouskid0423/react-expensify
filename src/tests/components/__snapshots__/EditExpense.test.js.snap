// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpense page behaviors. should successfully render a default snapshot. 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 7000,
        "createdAt": 172800000,
        "description": "Louisa Coffee",
        "id": 2,
        "note": "Cafe for working",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
